File: design/gen_icons.py
"""Rasterise the Branchwork mark into the icon files browsers ask for.

    .venv\\Scripts\\python design\\gen_icons.py

Writes into ``static/branchwork/img/``. The outputs are committed, so this
only needs re-running when the mark changes; Pillow is a dev dependency
(requirements-dev.txt) and is never imported by the app.

The geometry here is the same 32-unit grid as ``favicon.svg`` — the bold
version of the mark, not ``logo.svg``, because thin bars disappear at 16px.
Keep the two in step by hand: there is no SVG renderer in the loop, since
pulling in cairo to redraw three rectangles is a poor trade.
"""
import os

from PIL import Image, ImageDraw

OUT = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                   "static", "branchwork", "img")

GRID = 32                      # the viewBox of favicon.svg
RADIUS = 2.5
BARS = [  # x, y, w, h, fill — green, blue, red, as in theme.css
    (2, 3, 8, 26, "#3fcf63"),
    (12, 3, 8, 26, "#5b8def"),
    (22, 3, 8, 26, "#f0684e"),
]
# iOS ignores transparency and composites its own rounded mask over the
# square, so the touch icon gets the app background painted in.
APP_BG = "#15181d"
SUPERSAMPLE = 8


def render(size: int, background: str | None = None, mark: float = 1.0) -> Image.Image:
    """Square icon ``size`` px wide, the mark filling ``mark`` of the width."""
    canvas = size * SUPERSAMPLE
    image = Image.new("RGBA", (canvas, canvas), background or (0, 0, 0, 0))
    draw = ImageDraw.Draw(image)
    span = canvas * mark
    offset = (canvas - span) / 2
    unit = span / GRID
    for x, y, w, h, fill in BARS:
        draw.rounded_rectangle(
            [offset + x * unit, offset + y * unit,
             offset + (x + w) * unit, offset + (y + h) * unit],
            radius=RADIUS * unit, fill=fill)
    return image.resize((size, size), Image.LANCZOS)


def main() -> None:
    os.makedirs(OUT, exist_ok=True)

    for size in (16, 32):
        path = os.path.join(OUT, f"favicon-{size}.png")
        render(size).save(path)
        print("wrote", os.path.relpath(path, os.path.dirname(OUT)))

    # One .ico carrying three sizes, for browsers and Windows pinned sites
    # that still ignore the SVG.
    ico = os.path.join(OUT, "favicon.ico")
    render(64).save(ico, sizes=[(16, 16), (32, 32), (48, 48)])
    print("wrote", os.path.relpath(ico, os.path.dirname(OUT)))

    # 0.66 leaves the breathing room iOS home screens expect; flattened to
    # RGB because a transparent touch icon renders black on some versions.
    touch = os.path.join(OUT, "apple-touch-icon.png")
    render(180, background=APP_BG, mark=0.66).convert("RGB").save(touch)
    print("wrote", os.path.relpath(touch, os.path.dirname(OUT)))


if __name__ == "__main__":
    main()

File: design/gen_icons.mjs
// Build the icon files browsers ask for from static/branchwork/img/logo.svg.
// Outputs are committed, so this only runs when the logo changes:
//
//   npm install          (once; @resvg/resvg-js + png-to-ico, dev only)
//   node design/gen_icons.mjs
//
// resvg rather than cairo: it ships a prebuilt binary per platform, so this
// works on Windows without hunting for libcairo.
//
// logo.svg itself stays bare: the app's header draws it on the dark top bar.
// Every icon that lands in a browser tab or on a home screen gets the app's
// dark ground behind the mark instead, because the mark is drawn in white and
// vanishes on a light tab strip.
import { Resvg } from '@resvg/resvg-js';
import pngToIco from 'png-to-ico';
import { readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));
const IMG = join(ROOT, 'static', 'branchwork', 'img');
const LOGO = join(IMG, 'logo.svg');

const APP_BG = '#15181d';

// Tab icons: a rounded plate with the mark filling most of it, since at
// 16px every pixel of mark counts.
const FAVICON_UNITS = 64;
const FAVICON_INSET = 0.86;     // share of the plate the mark occupies
const FAVICON_RADIUS = 14;      // in FAVICON_UNITS

// iOS applies its own rounded mask, so the touch icon is a plain square with
// more room around the mark.
const TOUCH_SIZE = 180;
const TOUCH_INSET = 0.68;

const source = readFileSync(LOGO, 'utf8');

/** The logo's own viewBox, so the nested wrapper below scales correctly. */
function viewBox(svg) {
  const m = svg.match(/viewBox\s*=\s*"([^"]+)"/i);
  if (!m) throw new Error('logo.svg has no viewBox');
  return m[1];
}

/** Everything inside the outer <svg> element, prologue and doctype stripped. */
function innerMarkup(svg) {
  const open = svg.indexOf('<svg');
  const gt = svg.indexOf('>', open);
  const close = svg.lastIndexOf('</svg>');
  if (open < 0 || close < 0) throw new Error('logo.svg is not an svg');
  return svg.slice(gt + 1, close);
}

function render(svg, size) {
  return new Resvg(svg, { fitTo: { mode: 'width', value: size } }).render().asPng();
}

/** The mark centred on the app's ground, as a standalone SVG. */
function plated(units, inset, radius = 0) {
  const span = units * inset;
  const offset = (units - span) / 2;
  // xlink declared too: Affinity exports reference embedded images through it.
  return `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${units}" height="${units}" viewBox="0 0 ${units} ${units}">
  <rect width="${units}" height="${units}" rx="${radius}" fill="${APP_BG}"/>
  <svg x="${offset}" y="${offset}" width="${span}" height="${span}" viewBox="${viewBox(source)}">${innerMarkup(source)}</svg>
</svg>
`;
}

const favicon = plated(FAVICON_UNITS, FAVICON_INSET, FAVICON_RADIUS);
const written = [];

// The vector tab icon most browsers now prefer.
writeFileSync(join(IMG, 'favicon.svg'), favicon);
written.push('favicon.svg');

for (const size of [16, 32]) {
  writeFileSync(join(IMG, `favicon-${size}.png`), render(favicon, size));
  written.push(`favicon-${size}.png`);
}

writeFileSync(join(IMG, 'apple-touch-icon.png'), render(plated(TOUCH_SIZE, TOUCH_INSET), TOUCH_SIZE));
written.push('apple-touch-icon.png');

// One .ico carrying the small sizes, for browsers and Windows pinned sites
// that still ignore the SVG.
const ico = await pngToIco([16, 32, 48].map((s) => render(favicon, s)));
writeFileSync(join(IMG, 'favicon.ico'), ico);
written.push('favicon.ico');

console.log('wrote', written.join(', '), 'from logo.svg');
